backends/cloud/box/boxstorage.cpp

```diff
@@ -0,0 +1,299 @@
+#include "common/json.h"
+#include <curl/curl.h>
+#include "common/config-manager.h"
+
+namespace Cloud {
+namespace Box {
+
+char *BoxStorage::KEY = nullptr; //can't use CloudConfig there yet, loading it on instance creation/auth
+char *BoxStorage::SECRET = nullptr; //TODO: hide these secrets somehow
+
+void BoxStorage::loadKeyAndSecret() {
+	Common::String k = ConfMan.get("BOX_KEY", ConfMan.kCloudDomain);
+	KEY = new char[k.size() + 1];
+	memcpy(KEY, k.c_str(), k.size());
+	KEY[k.size()] = 0;
+
+	k = ConfMan.get("BOX_SECRET", ConfMan.kCloudDomain);
+	SECRET = new char[k.size() + 1];
+	memcpy(SECRET, k.c_str(), k.size());
+	SECRET[k.size()] = 0;
+}
+
+BoxStorage::BoxStorage(Common::String accessToken, Common::String refreshToken):
+	_token(accessToken), _refreshToken(refreshToken) {}
+
+BoxStorage::BoxStorage(Common::String code) {
+	getAccessToken(
+	    new Common::Callback<BoxStorage, BoolResponse>(this, &BoxStorage::codeFlowComplete),
+	    new Common::Callback<BoxStorage, Networking::ErrorResponse>(this, &BoxStorage::codeFlowFailed),
+	    code
+	);
+}
+
+BoxStorage::~BoxStorage() {}
+
+void BoxStorage::getAccessToken(BoolCallback callback, Networking::ErrorCallback errorCallback, Common::String code) {
+	if (!KEY || !SECRET) loadKeyAndSecret();
+	bool codeFlow = (code != "");
+
+	if (!codeFlow && _refreshToken == "") {
+		warning("BoxStorage: no refresh token available to get new access token.");
+		if (callback) (*callback)(BoolResponse(nullptr, false));
+		return;
+	}
+
+	Networking::JsonCallback innerCallback = new Common::CallbackBridge<BoxStorage, BoolResponse, Networking::JsonResponse>(this, &BoxStorage::tokenRefreshed, callback);
+	if (errorCallback == nullptr) errorCallback = getErrorPrintingCallback();
+	Networking::CurlJsonRequest *request = new Networking::CurlJsonRequest(innerCallback, errorCallback, "https://api.box.com/oauth2/token");
+	if (codeFlow) {
+		request->addPostField("grant_type=authorization_code");
+		request->addPostField("code=" + code);
+	} else {
+		request->addPostField("grant_type=refresh_token");
+		request->addPostField("refresh_token=" + _refreshToken);
+	}
+	request->addPostField("client_id=" + Common::String(KEY));
+	request->addPostField("client_secret=" + Common::String(SECRET));
+	/*
+	if (Cloud::CloudManager::couldUseLocalServer()) {
+	    request->addPostField("&redirect_uri=http%3A%2F%2Flocalhost%3A12345");
+	} else {
+	    request->addPostField("&redirect_uri=https%3A%2F%2Fwww.scummvm.org/c/code");
+	}
+	*/
+	addRequest(request);
+}
+
+void BoxStorage::tokenRefreshed(BoolCallback callback, Networking::JsonResponse response) {
+	Common::JSONValue *json = response.value;
+	if (!json) {
+		warning("BoxStorage: got NULL instead of JSON");
+		if (callback) (*callback)(BoolResponse(nullptr, false));
+		return;
+	}
+
+	Common::JSONObject result = json->asObject();
+	if (!result.contains("access_token") || !result.contains("refresh_token")) {
+		warning("Bad response, no token passed");
```

Comment:
Same thing, this need to have a prefix, so you could find it in the source code with ease.